matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -161,84 +166,201 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
                     + gpxList.size() + "). Correct format?");
         }
 
+        // filter the entries:
+        List<GPXEntry> filteredGPXEntries = filterGPXEntries(gpxList);
+        if (filteredGPXEntries.size() < 2) {
+            throw new IllegalStateException("Only " + filteredGPXEntries.size()
+                    + " filtered GPX entries (from " + gpxList.size()
+                    + "), but two or more are needed");
+        }
+
+        // now find each of the entries in the graph:
         final EdgeFilter edgeFilter = new DefaultEdgeFilter(algoOptions.getWeighting().getFlagEncoder());
 
-        // Compute all candidates first.
-        // TODO: Generate candidates on-the-fly within computeViterbiSequence() if this does not
-        // degrade performance.
-        final List<QueryResult> allCandidates = new ArrayList<>();
-        List<TimeStep<GPXExtension, GPXEntry, Path>> timeSteps = createTimeSteps(gpxList,
-                edgeFilter, allCandidates);
+        List<List<QueryResult>> queriesPerEntry = findGPXEntriesInGraph(filteredGPXEntries, edgeFilter);
 
-        if (allCandidates.size() < 2) {
-            throw new IllegalArgumentException("Too few matching coordinates ("
-                    + allCandidates.size() + "). Wrong region imported?");
-        }
-        if (timeSteps.size() < 2) {
-            throw new IllegalStateException("Coordinates produced too few time steps "
-                    + timeSteps.size() + ", gpxList:" + gpxList.size());
+        // now look up the entries up in the graph:
+        final QueryGraph queryGraph = new QueryGraph(routingGraph).setUseEdgeExplorerCache(true);
+        List<QueryResult> allQueryResults = new ArrayList<>();
+        for (List<QueryResult> qrs: queriesPerEntry)
+            allQueryResults.addAll(qrs);
+        queryGraph.lookup(allQueryResults);
+
+        logger.info("================= Query results =================");
```

Comment:
Ok. logger.debug sounds reasonable to me. Furthermore we should create a good debugging experience e.g. using something like the MiniGraphUI which requires also to move this project into the core to avoid circle deps.